### ISSUE TITLE
Allow starting GAP without GAPDoc, or ship GAPDoc with GAP

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#10]
Disclaimer: I like GAPDoc a lot, and this is not about getting rid of it, or not using it, or not installing it, or anything like that -- just making it easier to bootstrap a fresh GAP instance from source (see issue #5).

I think GAP should be a stand-alone project in the sense that it should be possible to download its source code, compile and then run it, without further extras. (I do not consider GMP and readline "extras" in this sense, as they are easily installable via package managers on most / all Unix systems, including most Linux distros and Mac OS X; and in any case, GAP works without readline, and by default builds its own GMP).

This is currently not possible, because GAP refuses to start if it cannot find GAPDoc. I see two ways around that:

1. Ship GAPDoc with GAP.
2. Allow starting GAP without GAPDoc.

Regarding the first option: We could include a copy of it in the repository. Or if Frank made a public git repository for GAPDoc somewhere, we could include it as a git submodule into the repository. I'd be happy with this/

As to the second option: Of course we should then print a big warning, as the documentation may not work. (Thinking about that, we probably should also print such a warning if GAPDoc is present, but the user has not compiled the manual...).

I just tested, and this minimal diff actually allowed me to start GAP without GAPDoc:

```diff
diff --git a/lib/package.gi b/lib/package.gi
index 375ebb2..4411a36 100644
--- a/lib/package.gi
+++ b/lib/package.gi
@@ -1832,6 +1832,10 @@ BindGlobal( "BANNER", false );
       fi;
     fi;

+    if not IsBound( GAPInfo.LoadPackageLevel ) then
+      GAPInfo.LoadPackageLevel:= 0;
+    fi;
+
     # If necessary then load the implementation part of the GAP library,
     # and the implementation parts of the packages loaded up to now.
     if not IsBound( GAPInfo.LibraryLoaded ) then
diff --git a/lib/system.g b/lib/system.g
index e4f5570..b11a880 100644
--- a/lib/system.g
+++ b/lib/system.g
@@ -28,7 +28,7 @@ BIND_GLOBAL( "GAPInfo", rec(

 # Without the needed packages, GAP does not start.
     Dependencies := rec(
-      NeededOtherPackages := [
+      SuggestedOtherPackages := [
         [ "gapdoc", ">= 1.2" ],
       ],
     ),
```

Some caveats apply:
1. There is an error in `transatl.g`, because it uses `SubstitutionSublist` which is supplied by GAPDoc. Solution: Add a copy of `SubstitutionSublist` (under a different name) to `transatl.g`
2. A big warning should be printed if GAPDoc is not present.
3. Tests should be run to see if other parts of GAP (or possibly packages) indirectly depend on GAPDoc functions. For code in GAP itself, I think we should then try to cut the dependency; for packages that use GAPDoc functions, I think those are fine, we just should suggest to their authors that they add GAPDoc to the `NeededOtherPackages` list.

Overall, this doesn't sound too difficult to me.

Thoughts?

**Comment by ChrisJefferson** [comment]
My personal preference would be to choose a small set of packages (I'd probably put GAPDoc and IO in that set) which we put in the GAP repository, and keep them up to date.

But I definitely agree we should either make it work without GAPDoc, or include it.

**Comment by frankluebeck** [comment]
On Wed, Mar 04, 2015 at 08:15:04AM -0800, Max Horn wrote:
> This is currently not possible, because GAP refuses to start if it
> cannot find GAPDoc. I see two ways around that:
> 
> 1. Ship GAPDoc with GAP.
> 2. Allow starting GAP without GAPDoc.

I strongly disagree with both of these possibilities. (And this has nothing
to do with the fact that I happen to be the author of GAPDoc.)

Ad 1. When we abandoned the old CVS repository for me and some others the
biggest step forward was that we got rid of the repositories of quite a few
independently developed (as well as quite a few unmaintained or outdated)
packages, these really didn't belong in a subdirectory of the core GAP 
repository. 

When you install the release version of GAP then GAPDoc IS shipped with GAP
(as well as > 100 other packages). This github repository does not (and
should not) provide an altenative for installing GAP. It is just the
repository for the core part of GAP, and this core part is not and 
should not be sufficient to run GAP as from the release version.

GAPDoc and many other packages are developed and released independently.
To run the development version of GAP you have, of course, to collect all the
components you want to use.

Ad 2. A while ago we have discussed and introduced dependencies of the core
part of GAP on packages. Dependencies between packages were common since
long, so that allowing such dependencies for the core part just makes the
core less special and more like one of the many packages. The GAPDoc package
was used as a prototype example to introduce such a dependency. De
facto, a minimal GAP also depends on the IO and Browse packages which
provide functionality to the user interface which many users like (and
most of them probably don't notice that these packages are involved).

We also wanted to encourage with this setup that more parts of the core GAP
are moved out into separate packages (which in my eyes has many advantages
on which I don't want to comment here).

-----

Ad the actual question how to run the development version of core GAP:

* install the release version of GAP (quite probably it is already 
  on your system)
* clone this (or rather your fork of this) repository to your machine
* within the clone add a link 'pkg' to a directory outside the cloned 
  tree
* in that external package directory add links to the packages in the
  release version
* if it seems necessary for whatever reason substitute some of these linked 
  packages with another version (e.g., just a copy of the released 
  package but compiled differently, or a clone of the repository
  of the dev-version of a package).

**Comment by alex-konovalov** [comment]
@frankluebeck : great, thanks all details! I suggest to include your "how to run the development version of core GAP" description into CONTRIBUTING.md. 

All, please be aware that the `hpcgap-default` branch still contains some version of GAPDoc which was around there in Autumn 2009. It has some minimal adjustments needed to use it with HPC-GAP. I hope that eventually it will be removed from this branch too.

(Somehow this discussion goes in parallel in this issue and #5, I've just posted a suggestion there regarding update/package manager).

**Comment by alex-konovalov** [comment]
@frankluebeck: but what about the other suggestion - to allow starting GAP without GAPDoc: do you remember why and when we have decided to make it a needed package?

**Comment by frankluebeck** [comment]
> but what about the other suggestion - to allow starting GAP without GAPDoc: do you 
> remember why and when we have decided to make it a needed package?

I hoped to have answered that in "Ad 2." above. The question is not how to remove the current dependency on GAPDoc (several workarounds were possible). The question is if it should be allowed that core GAP depends on packages (and so essentially becomes a package itself). In my opinion: yes, because otherwise it will be more difficult to modularize the current setup better.

I don't remember which release this dependency was introduced, but there was certainly a discussion about it.

**Comment by stevelinton** [comment]
Firstly, I agree with Frank that we do want to allow packages to be needed by the library and so required to start GAP. There’s lots of good functionality in the packages that we should be using in the library.

In the long term, this might go away in that the “core” system might just be a language interpreter with basically no mathematical capability that doesn’t depend on anything.

As a general principle, It would be good to avoid loading big complicated packages to get small “utility” functions, which certainly used to be what was going on with GAPDoc. In these cases packages could be split, or a small amount of code moved into the library, just to keep things simple.

There is a practical problem that it would be very nice to be able to use automated test tools on the GAP git repository without needing anything else. One solution to that would be to include a current packages archive (or perhaps the one from the last GAP release) in dev/ or somewhere. Not the source code or anything, just a zip either of all the packages or of the transitively-needed ones of the core. That could be updated from time-to-time painlessly. Another solution would be to have make (or make testing or ….) download them from a standard place, but that might burn a lot of time and network bandwidth for something like travis.

	Steve

**Comment by alex-konovalov** [comment]
Thanks for clarifications. I certainly agree that it's fine to have packages required by the core system, and I would consider GAP installation with missing documentation to be broken. However, I also think that @fingolfin is right that this creates some entry barriers for new contributors, which would be much less surprised to see a big warning about missing GAPDoc or not compiled manuals, than by the error message. Even adding a command line option to bypass loading of ALL packages, including required ones, will simplify Jenkins tests for compiler warnings - they compile GAP with `-Werror` option and fail in case of any warnings occurring, so they do not need GAPDoc there. 

Regarding the current packages archives, we such archives internally for Jenkins tests. We even have several versions of them (for the next update/minor/major release). They are named like `packages-2015_03_03-11_50_UTC.tar.gz`. I could easily place them on GAP ftp server with certain frequency, for example - just let me know what would be the preferred way to proceed.

**Comment by dimpase** [comment]
We should facilitate installing necessary non-OS dependencies, such as GAPDoc. One option is to use [git submodules](http://git-scm.com/book/en/v2/Git-Tools-Submodules) for these purposes. Is there a GAPDoc git repo from which one can clone it?

**Comment by alex-konovalov** [comment]
Thanks, but perhaps submodules is not the way to go for GAP packages at least for the following reasons:
* not all packages are hosted on GitHub and even less of them have git repositories.
* if a package will be a submodule, then `pkg` directory will became a part of the repository and can't be symlinked to `pkg` from the GAP installation.

**Comment by dimpase** [comment]
I did not mean to suggest that all the GAP packages should be git submodules. I finit it annoying that a non-automatic action is currently needed to get a running copy of GAP from the git repo, and I suppose I am not alone here.

One can put git submodules into a separate subdirectory, and symlink them from pkg/

As well, a github-hosted repo is not needed, it can be pulled from anywhere.

**Comment by frankluebeck** [comment]
On Mon, Mar 09, 2015 at 03:10:43AM -0700, Dmitrii Pasechnik wrote:
> We should facilitate installing necessary non-OS
> dependencies, such as GAPDoc. One option is to use [git
> submodules](http://git-scm.com/book/en/v2/Git-Tools-Submodules) for
> these purposes. Is there a GAPDoc git repo from which one can clone
> it?

No, as I already pointed out above, this is not what should be done.
The correct approach is to use the released versions of the packages you
need. You get them from here (either as part of the big repository and
also each package separately):
   http://gap-system.org/Releases/index.html

There is no reason why dev-versions of packages should be usable with the
dev-version of core-GAP at any time, isn't that exactly the point of
independent development? Only the released versions should work together.
And so there is no need to have access to the repository of a package.

I'm a bit surprised that this discussion is coming up now. Nothing has
changed by introducing this new git-repository, you just have to do the same 
as you had to do since the old CVS repository was abandoned.

Of course, I have no objections if someone invents a convenient script or
mechanism to make packages easily available for the repository version of 
core GAP.

**Comment by rbehrends** [comment]
I'd strongly advise against using Git submodules (for either GAPDoc or anything else). They (and their counterparts in other DVCSs [1], such as subrepos in Mercurial) generally produce nothing but problems. While the idea of incorporating another repository by reference sounds enticing and simple in theory, in practice it is anything but. Google, for example, specifically created its "repo" tool to mitigiate/avoid some of these issues in the context of Android development.

Branching and merging in particular can become a pain point; they reintroduce some of the problems that plagued CVS (because commits are no longer really atomic operations) and add some new issues on their own.

If you want to incorporate GAPDoc (or third-party projects) by reference rather than integrating it outright, then the best solution is to make this part of the build process, not version control. One example how this can be done is eru ([link](https://github.com/buildinspace/peru)). While I would not advocate for peru specifically (because of its dependency on Python 3.3 or later), the basic idea is simple and sound.

[1] Even externals in Subversion suffer from some of the same problems, though centralization mitigates them somewhat.

**Comment by dimpase** [comment]
The issues behind git submodules are vastly overblown. There are huge projects that use git submodules (e.g. coq, IIRC) quite happily.

**Comment by dimpase** [comment]
> Of course, I have no objections if someone invents a convenient script or mechanism to make packages easily available for the repository version of core GAP.

git subversion is such a mechanism. Still objections? 

> I'm a bit surprised that this discussion is coming up now. 

I have raised this issue before. But as people insisted on keeping discussions private, it was conveniently forgotten ("no conclusion was reached"). This is actually one reason I an strongly for keeping track of discussions, otherwise we end up talking about the same thing over and over again.

**Comment by stevelinton** [comment]
> Want to shop safely and privately online? Get DoNotTrackMe Premium.
> 
> On Mon, Mar 09, 2015 at 03:10:43AM -0700, Dmitrii Pasechnik wrote:
> > We should facilitate installing necessary non-OS
> > dependencies, such as GAPDoc.

Agreed.

> 
> There is no reason why dev-versions of packages should be usable with the
> dev-version of core-GAP at any time, isn't that exactly the point of
> independent development? Only the released versions should work together.
> And so there is no need to have access to the repository of a package.
> 
Also agreed

> 
> Of course, I have no objections if someone invents a convenient script or
> mechanism to make packages easily available for the repository version of 
> core GAP.
> 

and this  bit.

Making it easier or more automatic to get necessary released packages (or maybe all released packages)
into a development GAP checkout would be helpful and sensible. 

So the merits or otherwise of git submodules are moot.

	Steve

**Comment by markuspf** [comment]
How about introducing `make` targets such as `make bootstrap-minimal`, `make bootstrap-default`, and `make bootstrap-full`, and have this download GAPDoc, a small collection of packages, and the full collection of packages?

This would also interoperate with a "package manager" of some sort, should that ever materialise.

**Comment by alex-konovalov** [comment]
Just to reiterate: we already use internally combined archives of all packages, and I could make them available under some fixed names e.g. `packages-master-latest.tar.gz` and  `packages-stable-latest.tar.gz` in a fixed location e.g. `ftp://ftp.gap-system.org/pub/gap/gap4packagesfordevs` so the bootstrapping script could look for them there. I would be able to update them periodically.

**Comment by alex-konovalov** [comment]
We agreed that I will be publishing under some fixed URLs the following archives:
- selection of all packages that were released with the latest stable release
- selection of all packages that would be release with the next minor release
- selection of packages that would be released with the next major release.

As a matter of fact, doing this now would result in three different archives, so having all three is completely sensible. I am assigning this issue to myself until this will be done.

**Comment by alex-konovalov** [comment]
@markuspf and @ChrisJefferson, 
here there are URLs:
* `ftp://ftp.gap-system.org/pub/gap/gap4pkgs/packages-latest.tar.gz` - selection of all packages that were released with the latest stable release
* `ftp://ftp.gap-system.org/pub/gap/gap4pkgs/packages-stable.tar.gz` - selection of all packages that would be release with the next minor release
* `ftp://ftp.gap-system.org/pub/gap/gap4pkgs/packages-master.tar.gz` - selection of packages that would be released with the next major release

I will be updating these periodically (`packages-latest` after each minor release, the other two after package updates were picked up and passed our weekly testing cycle to provide a reliable selection of packages for testing).

Please feel free to close this issue if you think that this resolves the issue.

**Comment by fingolfin** [comment]
Frank expressed surprise that this discussion comes up now. Well, one thing that changed is that the repository is now accessible by the public -- in the past, it was only available to a private club of insiders, so it being non-trivial to use was not as much of an issue...

Anyway,  providing some scripts (or at least some useful concise documentation) for getting a usable GAP out of the source repository would indeed also solve the problem. I.e. what to do after having done a clone and compiled in order to get a working GAP with sufficiently many packages (i.e. at least GAPDoc) as to be usable.

So, is anybody planing to work on that?

In the meantime, I still think that including a copy of the latest stable versions of any required GAP packages (which, currently, means only GAPDoc), would be the simplest and pragmatic solution to this. It could be reverted again once we have a better solution in place. And quite frankly, I am surprised that doing something like this is so controversial, maybe becaues I so far have not quite understood what the drawbacks of that would be (of course I am assuming these copies would not be modified, except for updates). 

We shoudl also remove or augment the following somewhat misleading part of the README:
```
If everything goes well, you should be able to start GAP by executing
    # sh bin/gap.sh
```
We should at least point out that yes, it is normal that this leads to this output:
```
 ┌───────┐   GAP, Version 4.dev of today (free software, GPL)
 │  GAP  │   http://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin12.6.0-gcc-default64
 Libs used:  gmp, readline
 Loading the library and packages ...
#I  GAP: needed package gapdoc cannot be loaded
Error, failed to load needed package `gapdoc' (version >= 1.2) called from
func(  ); on line 143 of file /Users/mhorn/Projekte/GAP/gap.github/lib/system.g called from
<function "CallAndInstallPostRestore">( <arguments> )
 called from read-eval loop at line 899 of /Users/mhorn/Projekte/GAP/gap.github/lib/init.g
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk>
```

Of course better wold be to provide proper bootstrapping instruction (see also issue #5).

**Comment by alex-konovalov** [comment]
Following discussion with @ChrisJefferson and @fingolfin, new names of the package archives from `ftp://ftp.gap-system.org/pub/gap/gap4pkgs/` have the format `packages-<branch/tag>.tar.gz`. At the moment, there are three of them there:
* `packages-master.tar.gz`	
* `packages-stable-4.7.tar.gz`  
* `packages-v4.7.7.tar.gz`

**Comment by fingolfin** [comment]
I'll try to improve the README, including pointers to these archives

**Comment by alex-konovalov** [comment]
I will try to document this elsewhere, since this information would be hard to dig out of issues in the future. But for now, since it's relevant to this discussion: asked by @ChrisJefferson, I've added three more archives to facilitate automated testing: `packages-required-stable-v4.7.7.tar.gz`, `packages-required-master.tar.gz` and `packages-required-stable-4.7.tar.gz`. At the moment, they contain only GAPDoc package. All archives are available by FTP and HTTP, here there are http links:

All packages:
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-master.tar.gz
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-stable-4.7.tar.gz
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-v4.7.7.tar.gz

Required packages:
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-required-stable-v4.7.7.tar.gz
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-required-master.tar.gz
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-required-stable-4.7.tar.gz

**Comment by alex-konovalov** [comment]
I'd like to summarise the discussion and suggest a list of items to be done to close this issue:

1) Clarify in `README.md` and `CONTRIBUTING.md` that it's normal if the GAP core system build from this repository can not be started because of missing GAPDoc, as suggested by @fingolfin, and incorporate suggestions on how to run the development version of core GAP system from a comment by @frankluebeck on March 4th

2) Document somewhere fixed URLs of merged package archives as listed by me in the previous comment.

3) Implement make targets (suggested by @markuspf on March 9th) possibly named as `make bootstrap-minimal` and `make bootstrap`, which will use that archives to populate the `pkg` directory and compile as many packages as possible.

Are there any objections/additions/comments?

**Comment by alex-konovalov** [comment]
FYI: Because of GAP 4.7.8 release, I've now posted archives
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-v4.7.8.tar.gz
* http://www.gap-system.org/pub/gap/gap4pkgs/packages-required-stable-v4.7.8.tar.gz

with all packages and required packages respectively.

**Comment by markuspf** [comment]
I will implement a couple of Makefile targets that allow bootstrapping of the repository shortly.

**Comment by markuspf** [comment]
The merged pull-request #183 addresses this issue for now.
